File: front-blazor/Services/Services/MenuProvider.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Front.Services.Models;
using Tiny.Framework.Utility;

namespace Front.Services.Services
{
    internal sealed class MenuProvider :
        IMenuProvider
    {
        public MenuProvider(ICatalogueFilter filter)
        {
            It.IsNull(filter)
                .AsGuard<ArgumentNullException>(nameof(filter));

            Filter = filter;
        }

        public ICatalogueFilter Filter { get; set; }

        public IMenuNode CreateBranch(string name)
        {
            return new MenuNode
            {
                Name = name
            };
        }

        public async Task<IMenuNode> GetPagesMenu(string subjectTitle, ICatalogue fromCatalogue) =>
            await GetSimpleMenu(subjectTitle, MenuCategories.Pages, fromCatalogue);

        public async Task<IMenuNode> GetExternalMenu(string subjectTitle, ICatalogue fromCatalogue) =>
            await GetSimpleMenu(subjectTitle, MenuCategories.External, fromCatalogue);

        public async Task<IMenuNode> GetArticlesMenu(string subjectTitle, ICatalogue fromCatalogue)
        {
            var menu = CreateBranch(subjectTitle);
            var yearBranch = default(IMenuNode);
            var monthBranch = default(IMenuNode);

            (await GetFilteredArticles(MenuCategories.Articles, fromCatalogue))
                .OrderByDescending(x => x.PublicationDate)
                .AsSafeReadOnlyList()
                .ForEach(post =>
                {
                    var node = CreateLeaf(post);

                    yearBranch = menu.Children.FirstOrDefault(x => x.Name == node.Year);
                    if (yearBranch is null)
                    {
                        yearBranch = CreateBranch(node.Year);
                        menu.Children.Add(yearBranch);
                    }

                    monthBranch = yearBranch.Children.FirstOrDefault(x => x.Name == node.Month);
                    if (monthBranch is null)
                    {
                        monthBranch = CreateBranch(node.Month);
                        yearBranch.Children.Add(monthBranch);
                    }

                    monthBranch.AddNode(node);
                    yearBranch.IncrementCount();
                });

            return menu;
        }

        internal IMenuNode CreateLeaf(IPostedArticle article)
        {
            return new MenuNode
            {
                Name = article.Title,
                Topic = article.Topic != MenuCategories.External ? article.Topic : article.Path,
                Year = article.Year,
                Month = article.Month,
                Article = article.Article
            };
        }

        internal async Task<IEnumerable<IPostedArticle>> GetFilteredArticles(string filter, ICatalogue fromCatalogue) =>
            await Filter.FindArticles(filter, fromCatalogue);

        internal async Task<IMenuNode> GetSimpleMenu(string subjectTitle, string filter, ICatalogue fromCatalogue)
        {
            var menu = CreateBranch(subjectTitle);

            (await GetFilteredArticles(filter, fromCatalogue))
                .AsSafeReadOnlyList()
                .ForEach(post =>
                {
                    var node = CreateLeaf(post);
                    menu.Children.Add(node);
                });

            return menu;
        }
    }
}

File: front/src/app/services/menu.provider.ts
import { Injectable } from '@angular/core';

import { MenuNode } from '../models/menu.node';
import { menuCategories } from '../models/menu.categories'
import { PostedArticle } from '../models/posted.article';
import { ICatalogue } from '../models/catalogue';
import { CatalogueFilter } from './catalogue.filter.provider';

@Injectable()
export class MenuProvider {

    constructor(
        private filter: CatalogueFilter)
    {
    }

    public createBranch(name: string): MenuNode {
        return {
            name: name,
            count: 0,
            children: []
        };        
    }

    private createLeaf(article: PostedArticle): MenuNode {
        return {
            name: article.title,
            topic: article.topic !== menuCategories.external ? article.topic : article.path,
            year: article.year,
            month: article.month,
            article: article.article,
            count: 0
        };        
    }

    private getFilteredArticles(filter: string, fromCatalogue: ICatalogue): PostedArticle[] {
        return this.filter.findArticles(filter, fromCatalogue);
    }

    private getSimpleMenu(subjectTitle: string, filter: string, fromCatalogue: ICatalogue): MenuNode {
        var menu = this.createBranch(subjectTitle);

        this.getFilteredArticles(filter, fromCatalogue)
            .forEach(post => {
                var node = this.createLeaf(post);
                menu.children.push(node);
            });

        return menu;
    }

    public getPagesMenu(subjectTitle: string, fromCatalogue: ICatalogue): MenuNode {
        return this.getSimpleMenu(subjectTitle, menuCategories.pages, fromCatalogue);
    }

    public getExternalMenu(subjectTitle: string, fromCatalogue: ICatalogue): MenuNode {
        return this.getSimpleMenu(subjectTitle, menuCategories.external, fromCatalogue);
    }

    public getArticlesMenu(subjectTitle: string, fromCatalogue: ICatalogue): MenuNode {
        var menu = this.createBranch(subjectTitle);
        var yearBranch: MenuNode;
        var monthBranch: MenuNode;

        this.getFilteredArticles(menuCategories.articles, fromCatalogue)
            .sort((x, y) => 0 - (x.publicationDate > y.publicationDate ? 1 : -1))
            .forEach(post => {
                var node = this.createLeaf(post);

                yearBranch = menu.children.find(x => x.name == node.year);
                if(!yearBranch) {
                    yearBranch = this.createBranch(node.year);
                    menu.children.push(yearBranch);
                }

                monthBranch = yearBranch.children.find(x => x.name == node.month);
                if(!monthBranch) {
                    monthBranch = this.createBranch(node.month);
                    yearBranch.children.push(monthBranch);
                }

                monthBranch.children.push(node);
                monthBranch.count++;
                yearBranch.count++;
            });

        return menu;
    }
}
